Add optional CI workflow for clippy and cross-compilation tests

hpwxf opened the pull request:
This PR includes a additional CI workflow for testing git2-rs for various targets.
```
          # - { os: ubuntu-latest  , target: arm-unknown-linux-gnueabihf , use-cross: use-cross }
          - { os: ubuntu-latest  , target: aarch64-unknown-linux-gnu   , use-cross: use-cross }
          - { os: ubuntu-latest  , target: aarch64-unknown-linux-musl  , use-cross: use-cross }
          - { os: ubuntu-latest  , target: i686-unknown-linux-gnu      , use-cross: use-cross }
          - { os: ubuntu-latest  , target: i686-unknown-linux-musl     , use-cross: use-cross }
          - { os: ubuntu-latest  , target: x86_64-unknown-linux-gnu    , use-cross: use-cross }
          - { os: ubuntu-latest  , target: x86_64-unknown-linux-musl   , use-cross: use-cross }
          - { os: macos-latest   , target: x86_64-apple-darwin }
          # - { os: windows-latest , target: i686-pc-windows-gnu }
          - { os: windows-latest , target: i686-pc-windows-msvc }
          - { os: windows-latest , target: x86_64-pc-windows-gnu }
          - { os: windows-latest , target: x86_64-pc-windows-msvc }
```
I found test or compilation errors in the targets as comments.

This CI workflow also include a clippy check for linux/mac/windows (due to specific sections for windows -- mac target could not be useful since there is no specific mac section)

Activating clippy has produced many warnings (promoted as errors in this CI workflow). I've fixed all of them (and sometimes I add an `allow(clippy:xxx)` pragmas:
* `clippy::new_without_default`
* `clippy::unnecessary_cast`
* `clippy::too_many_arguments`
* `clippy::manual_non_exhaustive`
* `clippy::should_implement_trait`
* `clippy::transmute_ptr_to_ptr`

I note 2 special cases of maybe useless "unsafe" function. I put them in a separate commit.

A coverage section is ready to use in the same workflow.

alexcrichton commented:
Thanks for the PR! This is a lot of code to review though and personally I would prefer to not take on a whole lot of CI configuration for this (I personally do not like to run clippy in CI). Could this be minimized into just the clippy changes?

hpwxf commented:
Ok; I understand.
I move the CI to another branch (my idea will next be to investigate why there are problems on arm-unknown-linux-gnueabihf and i686-pc-windows-gnu).

To simplify the review I also split the changes into 4 commits:
* shorthand initialization warnings
* clippy directives to disable some warnings (because seems to be easier to understand like this)
  (eg: `allow(clippy::new_without_default)`, `allow(clippy::unnecessary_cast)`...)
* all the others changes (except 2 in the next commit)
* 2 `unsafe` remove
If you prefer, I can also split these changes into distinct PRs.

alexcrichton commented:
Thanks, but I generally don't agree with many of the changes here. It's unfortunately pretty exhausting to deal with clippy PRs across so many projects. I don't really have the time/energy to review all sorts of clippy prs all over the place taking various stylistic actions that I don't agree with personally.

I'll leave this though if someone else would like to review.

hpwxf commented:
No problem. I understand. Maybe clippy's advice isn't that helpful. It was just a suggestion (maybe not a good one). You can close this PR, if you prefer.
I did this in passing before trying to figure out the runtime and compilation issues on the two target architectures (`arm-unknown-linux-gnueabihf` and `i686-pc-windows-gnu`). Instead, I will focus on my original goal.
Thanks for your time.

ehuss commented:
I'm going to go ahead and close this out since it is quite old.
